_posts/2025-11-06-sealing.markdown

```diff
@@ -0,0 +1,202 @@
+This lets you pass untrusted code a pointer to something that they can pass back but can't tamper with.
+
+# Sealing gives type safety
+
+The object type in a CHERI capability doesn't necessarily have to have a 1:1 mapping to a language-level type.
+It's quite common for a set of types to have some kind of internal type discriminator.
+For example, in our [ASPLOS 2017 CHERI JNI paper](https://www.cl.cam.ac.uk/research/security/ctsrd/pdfs/201704-asplos-cherijni.pdf), we used three otypes for everything passed from the JVM to native code.
+One each for field-ID and method-ID structures, and one for all Java objects.
+Each Java object starts with a pointer to its class, so we didn't need one otype for each Java class type, just one for all Java objects.
+The virtualised sealing mechanism in CHERIoT uses this same approach to multiplex a huge number of possible object types onto two hardware otypes (one for statically allocated objects and one for dynamic, so the memory allocator is not in the TCB for statically allocated sealed objects).
+More on this later.
+
+If you couple a sealing key type with a language-level type (or family of self-disambiguating types), then you can build type safety that works in the presence of an attacker.
+It's common in C/C++ to expose an *opaque type* at API boundaries.
+This is usually implemented as a pointer to forward-declared structure type.
+C and C++ will not let you dereference this pointer unless you cast it to another type.
+CHERI sealing lets you enforce the no-dereference rule, even if malicious code does cast it to some other type.
+This means that you can hand sealed capabilities to other compartments and they must treat them as opaque types.
+When you get them back, you have a lightweight check that they really are the type that you expect.
+
+After the 2015 paper, we separated the permit-seal and permit-unseal permissions.
+In most CHERIoT use cases, the entity that can seal and unseal pointers with a particular otype is the same.
+This isn't universal.
+C++ provided one of our original use cases for separating them.
+If you seal C++ vtables with a well-known otype, and make the permit-unseal capability for it available anywhere, then you can have a C++ ABI where only the loader can forge vtables.
+This makes code reuse attacks harder.
+
+There are other situations where an object type is used for *integrity* but not *confidentiality*.
+It is an attestation that some software trusted to seal with a particular otype has done so, even if anyone is then able to unseal the result.
+The sentry mechanism (discussed later) is a variant of this idea.
+
+# Virtualised sealing shares hardware object types
+
+Morello had a 15-bit object-type field, which is sufficient for a lot of things.
+When we scaled CHERI down to 32-bit systems for CHERIoT, we ended up with only three bits of space.
+Of these, the zero value means not-sealed, so there were only seven values available.
```

Comment:
"And even with the insight that executable pointers and data pointers wanted disjoint types, that's still only 14." or somesuch?  Maybe skip it until sentries.